src/Workspaces/Core/Portable/Storage/CloudCache/CloudCachePersistentStorage.cs

```diff
@@ -0,0 +1,211 @@
+                pipe.Reader.AdvanceTo(readResult.Buffer.Start, readResult.Buffer.End);
+
+                if (readResult.IsCompleted)
+                    break;
+            }
+
+            return pipe.Reader.AsStream();
+        }
+
+        public override Task<bool> WriteStreamAsync(string name, Stream stream, Checksum? checksum, CancellationToken cancellationToken)
+            => WriteStreamAsync(name, stream, checksum, s_solutionKey, cancellationToken);
+
+        public override Task<bool> WriteStreamAsync(ProjectKey projectKey, string name, Stream stream, Checksum? checksum, CancellationToken cancellationToken)
+            => WriteStreamAsync(name, stream, checksum, GetContainerKey(projectKey), cancellationToken);
+
+        public override Task<bool> WriteStreamAsync(DocumentKey documentKey, string name, Stream stream, Checksum? checksum, CancellationToken cancellationToken)
+            => WriteStreamAsync(name, stream, checksum, GetContainerKey(documentKey), cancellationToken);
+
+        private async Task<bool> WriteStreamAsync(string name, Stream stream, Checksum? checksum, CloudCacheContainerKey? containerKey, CancellationToken cancellationToken)
+        {
+            // If we failed to get a container key (for example, because teh client is referencing a file not under the
+            // solution folder) then we can't proceed.
+            if (containerKey == null)
+                return _mustSucceed ? throw new InvalidOperationException("Could not create container key") : false;
+
+            if (checksum == null)
+            {
+                return await WriteStreamAsync(new CloudCacheItemKey(containerKey.Value, name), stream, cancellationToken).ConfigureAwait(false);
+            }
+            else
+            {
+                using var bytes = s_byteArrayPool.GetPooledObject();
+                checksum.WriteTo(bytes.Object);
+
+                return await WriteStreamAsync(new CloudCacheItemKey(containerKey.Value, name, bytes.Object.AsMemory()), stream, cancellationToken).ConfigureAwait(false);
+            }
+        }
+
+        private async Task<bool> WriteStreamAsync(CloudCacheItemKey key, Stream stream, CancellationToken cancellationToken)
+        {
+            // From: https://dev.azure.com/devdiv/DevDiv/_wiki/wikis/DevDiv.wiki/12664/VS-Cache-Service
+            var pipe = new Pipe();
+            var addItemTask = _cacheService.SetItemAsync(key, pipe.Reader, shareable: false, cancellationToken);
+            try
+            {
+                await stream.CopyToAsync(pipe.Writer, cancellationToken).ConfigureAwait(false);
+                await pipe.Writer.CompleteAsync().ConfigureAwait(false);
+            }
+            catch (Exception ex)
+            {
+                await pipe.Writer.CompleteAsync(ex).ConfigureAwait(false);
+                throw;
+            }
+
+            await addItemTask.ConfigureAwait(false);
```

Comment:
Seeing as you already have a stream with the content, why not reduce this all with the [`AsPipeReader` extension method](https://apisof.net/catalog/61be72ab-a3cf-8add-af46-d8365de21083) to just:

```cs
await _cacheService.SetItemAsync(key, stream.AsPipeReader(), shareable: false, cancellationToken);
```

Reply:
This whole method can be replaced now if you update RPC.Contracts to the latest version, which adds extension methods for SetItemAsync and TryGetItemAsync that work with Stream directly.
You should perhaps _not_ use the `TryGetItemAsync` extension method however, since you have the special requirement to buffer the result before turning it into a stream.

Reply:
done.